tests/testlib/s2n_kem_fuzz_testlib.c

```diff
@@ -17,95 +17,82 @@
 #include "utils/s2n_safety.h"
 #include "tls/s2n_kem.h"
 #include "tests/testlib/s2n_nist_kats.h"
+#include "pq-crypto/s2n_pq.h"
 
 int s2n_kem_recv_ciphertext_fuzz_test_init(const char *kat_file_path, struct s2n_kem_params *kem_params) {
     notnull_check(kat_file_path);
     notnull_check(kem_params);
-
-#if defined(S2N_NO_PQ)
-    return S2N_FAILURE;
-#else
     notnull_check(kem_params->kem);
 
     GUARD(s2n_alloc(&kem_params->private_key, kem_params->kem->private_key_length));
-
     FILE *kat_file = fopen(kat_file_path, "r");
     notnull_check(kat_file);
     GUARD(ReadHex(kat_file, kem_params->private_key.data, kem_params->kem->private_key_length, "sk = "));
     fclose(kat_file);
 
     return S2N_SUCCESS;
-#endif
 }
 
 int s2n_kem_recv_ciphertext_fuzz_test(const uint8_t *buf, size_t len, struct s2n_kem_params *kem_params) {
     notnull_check(buf);
     notnull_check(kem_params);
-
-#if defined(S2N_NO_PQ)
-    return S2N_FAILURE;
-#else
     notnull_check(kem_params->kem);
 
-    /* Because of the way BIKE1_L1_R1's decapsulation function is written, this test will not work for that KEM. */
-    ENSURE_POSIX(kem_params->kem != &s2n_bike1_l1_r1, S2N_ERR_KEM_UNSUPPORTED_PARAMS);
-
-    struct s2n_stuffer ciphertext = { 0 };
-    GUARD(s2n_stuffer_growable_alloc(&ciphertext, 8192));
+    DEFER_CLEANUP(struct s2n_stuffer ciphertext = { 0 }, s2n_stuffer_free);
+    GUARD(s2n_stuffer_alloc(&ciphertext, len));
     GUARD(s2n_stuffer_write_bytes(&ciphertext, buf, len));
 
-    /* Don't GUARD here; this will probably fail. */
-    s2n_kem_recv_ciphertext(&ciphertext, kem_params);
-    /* The PQ KEM functions are written in such a way that kem->decapsulate should
-     * never fail (except for BIKE1_L1_R1), even if the ciphertext is not valid. So,
-     * we check it with GUARD. */
-    if (kem_params->shared_secret.allocated == 0) {
-        /* If s2n_kem_recv_ciphertext failed, this probably did not get allocated. */
-        GUARD(s2n_alloc(&kem_params->shared_secret, kem_params->kem->shared_secret_key_length));
+    /* Don't GUARD here; this will probably fail. We should fuzz this
+     * top-level function regardless of whether or not PQ is enabled. */
+    int recv_ciphertext_ret = s2n_kem_recv_ciphertext(&ciphertext, kem_params);
+
+    /* The recv_ciphertext() function may fail, but as long as PQ is enabled,
+     * it should never fail due to S2N_ERR_PQ_CRYPTO (the only exception is
+     * BIKE1L1R1, which may fail due to a PQ_CRYPTO error because of the way
+     * that KEM's decaps function is written). */
+    if (s2n_pq_is_enabled() && recv_ciphertext_ret != S2N_SUCCESS && kem_params->kem != &s2n_bike1_l1_r1) {
+            ne_check(s2n_errno, S2N_ERR_PQ_CRYPTO);
     }
```

Comment:
Why might this fail? Are you finding other, non-PQ issues that we need to fix?

Reply:
That comment is a bit unclear; I will clarify it.

* `recv_ciphertext()` parses the bytes from the handshake (for all KEMs), so it will reject input that isn't formatted correctly (realistically, I think it will only fail if the input ciphertext [isn't properly length-encoded ](https://github.com/awslabs/s2n/blob/5c6801b33ce5001b3db5c989e00938fc6fed5dcd/tls/s2n_kem.c#L470-L471)).

* The BIKE1L1R1 KEM is a bit of a special case. All of the other KEMs in s2n (including BIKE round 2) make a guarantee that even if the inputs (private key and ciphertext) to the decaps function are garbage, the execution of the decaps function will still succeed, but the output (plaintext) will be garbage. BIKE1L1R1 doesn't make this guarantee, and so providing garbage inputs to its decaps function means that it will likely just outright fail and return non-zero.